### DISCUSSION TITLE
How to port litematica

### DISCUSSION THREAD

**Discussion by ZacSharp** [maruohon/litematica#318] [Q&A] [answered]
I'm now two months past the point where I actually needed Litematica and since my first attempt at porting it myself failed and my own approach for a schematic mod doesn't render in 1.16 I guess I'll ask here.
How do I port Litematica either from fabric to Forge or from 1.12 to 1.16?
I know that I have to adjust the buildscript, change some classnames and work around changed/missing functionality on the new target platform, but I'm hoping for some more detailed help especially for the buildscript and functionality changes (I don't know much about them).
Some way to change the mappings without searching and (bulk-)replacing everything by hand would be nice nonetheless.

**Comment by maruohon** [comment, marked as answer]
I can't just give you clear instructions on everything, as I would need to do the port myself first to even be able to know everything that needs to change. The port would by far be easier to do from the 1.16 Fabric version, as it has already adapted to all the vanilla changes since 1.12. And now that Forge has "support" for (or at least ships with) Mixin, the port might be almost entirely just class, field and method renames from Yarn to MCP. You would need to replace the main mod class and the metadata files the mod loader uses to load the mod obviously, as those are different in Fabric vs. Forge.

For build script changes compared to the Fabric versions, you could for example diff/look at the build scripts of the 1.15 port of malilib. I didn't use Mixins in those ports because it was not shipped with Forge back then, so I don't actually know yet how to set up Mixins under the current Forge versions.

You should start from porting malilib first, and once you have that running and working (there isn't much visible beyond the config menu `A + C`), then you can start on Litematica. Beyond this you are on your own if you are attempting a port. I don't have time to try to debug things for your port, especially since it would be impossible to do without actually starting to work on the port myself to see what is what. I am going to make my own port in any case once I get the new updates ready on Fabric first.

So although I realize that it's frustrating for people to wait on the Forge ports to happen (believe me, it's also frustrating on my part for not managing to make faster progress on things), I don't want to port the current versions directly, as that would be wasting more of my modding time on a temporary solution that will be discarded once the new code is ready. The current development code is getting so many massive changes that it may be unfeasible to try to merge the development code from 1.12.2 on top of the current 1.13.2+ versions, it may end up being easier just doing each of the MC version ports again on top of the new 1.12.2 code. A further complication for the merge are the mapping differences between versions and mod loaders.

The current progress on the development code on 1.12.2 is that malilib itself is getting closer to having all the planned features and additions. Thus I should hopefully get to working on the improvements and clean-ups and rewrites and new features for the other mods soon-ish.

**Reply by ZacSharp** [reply]
Done, code is on the forrge_1.16.5 branches of my forks.
Malilib was mainly a version bump and a bunch of MatrixStacks.
Litematica was a little more, but I managed to avoid renaming all methods by [this](https://www.tterrag.com/yarn2mcp/) (still had to do the classes). The two things that surprised me were MaLiLibs  WorldEventHandler#onWorldLoadPost never being called and Forge calling model data functions on arbitrary client worlds throwing an error if it is not the current client world ("model data can only be used on the current client world")

Would you be interested in building for fabric from mcp named sources (or whatever mappings your liteloader branch uses), like done in https://github.com/cabaletta/baritone/pull/2401? I could try to do it.
EDIT: Have the feeling that I should clarify this a little:
I would try to build for fabric from one of the 1.13+ forge branches, to hopefully create something that can be used to port from liteloader 1.12.2 to newer mc versions without changing the mappings. Ideally I'd like to try to build forge versions from all current fabric branches, but I don't know how to add mappings for the class names (didn't find any mappings with class names in the required format).

**Reply by maruohon** [reply]
I don't want to build for Fabric using MCP names, even if it's possible. I would rather develop for Forge using Yarn names, as I prefer the Yarn names in pretty much all the cases where the names differ between the two. I also tried yarn2mcp at one point, but the environment somehow managed to corrupt itself pretty much immediately, so I didn't really get anywhere with that experiment at that time...

But basically I want to keep the Fabric versions as the primary versions on each MC version, as Fabric is what I use/prefer myself when developing and playing (if that ever happens). The Forge versions would thus always be ports from the Fabric versions. So either using yarn2mcp or somehow cooking up some custom mappings and feeding them to Loom's `migrateMappings` task to rename between the two is probably how I'm going to do things. I think `migrateMappings` doesn't handle mixin classes though unfortunately, so the renaming probably can't be *fully* automated, for now at least.

Also like I mentioned in the earlier message, there are massive refactors and clean-up happening in the 1.12.2 LiteLoader branches atm, so I'm pretty sure the next ports to 1.13.2+ will be made basically from scratch, looking at the old diff between the branches to easily see what areas of the code need changes between the respective MC versions. I *think* that's probably going to be easier that trying a normal merge and resolving all the thousands of conflicts in the hundreds of different classes.

**Reply by ZacSharp** [reply]
Ok, so I won't bother you again unless I find a way to develop fabric and forge with yarn on the same branch.

To me [MercuryMixin](https://github.com/CadixDev/MercuryMixin) looks like it could do it fully automatically, but I haven't looked at more than it's readme yet and out of the remapping programs I actually tried none worked, so my hopes aren't high.

I know you are doing massive and likely completely diff-incompatible changes and my idea was to end up with a build script that could be copied to the new port with minimal adjustments (No litematica or malilib java code involved, so not affected by your refactors).

**Reply by maruohon** [reply]
> Ok, so I won't bother you again unless I find a way to develop fabric and forge with yarn on the same branch.

I'm perfectly fine discussing these things here, I hope I didn't come off rude or uninterested in the Forge ports in general, as that was not my intention. I just don't want to spend any time working on the current 1.16 (or really the 1.13.2+) code base, as I personally don't like that code with its many limitations (from malilib's side especially), and especially the user experience and user interface (mostly the config menu layout) those current mod versions offer.

So at the moment I want to focus on the rewrites and improvements in the main development branch, which will bring numerous major improvements and new capabilities (via malilib) to many areas of the code and especially the (config) menus. And also as part of these upcoming updates I'm trying to work on a bunch of the most often requested features and improvements to all of the client mods. So once I get these updates ready, it should considerably improve the overall user experience and also the feature set of the mods. That's also another reason why I don't want to port the current versions to Forge, as that will expand the user base further, and I would like to offer those new users the better user experience the upcoming versions will offer. I'm just somewhat frustrated how slowly I seem to get stuff done and things crossed off my todo list, so these updates are just dragging along timeline-wise. (The original estimates I gave to people around last summer were by the end of the year... yeah... and stuff is still nowhere near finished. malilib is now getting pretty close to being feature-complete, just one new system to add and then some further clean-up and a few bug fixes etc. left on the todo list.)

So once these updates are done for 1.12.2, I can finally port all of this code to the 1.13.2+ versions again, as the last merge from 1.12.2 to 1.13.2+ happened all the way back in June 2019. Since then the 1.12.2 versions have been slowly moving forward, while the 1.13.2+ versions have been stuck more or less there feature-wise. There have been just a number of smaller features cherry-picked or directly implemented "temporarily" directly to the 1.16 version for example, that are intended to go to 1.12.2 later on and then be merged/ported forward via the "main branches" from 1.12.2. This is also the main reason for the current "temp" branches that exist for some of the mods on 1.16, as I probably will end up throwing away those "temporarily ported or added features" branches once I do the big merge/update from 1.12.2, as it won't apply cleanly on top of those branches.
